src/ipc-win32.cc

```diff
@@ -0,0 +1,251 @@
+
+#include "util.h"
+#include "version.h"
+
+// Win32 implementation of IPC for requesting builds from a persistent build
+// server. We use Win32 named pipes for communication and synchronization.
+// Because Win32 doesn't support fork(), when spawning a server we have to use
+// CreateProcess to start a new process from the beginning of main(). We signal
+// to the new process that it should be a build server by creating a Win32 named
+// event object that it can set once it creates the named pipe.
+
+static const int max_message_size = 1024 * 256;
+
+string GetCwd() {
+  int size = GetCurrentDirectory(0, NULL);
+  vector<char> cwd(size);
+  GetCurrentDirectory(size, cwd.data());
+  return cwd.data();
+}
+
+// Returns a string containing all of the state that can affect a build, such as
+// ninja version and arguments. The server checks to make sure this matches the
+// client before building.
+string GetStateString(int argc, char **argv) {
+  string state;
+  // If the current working directory is longer than 246 characters then it will
+  // be truncated in the pipe name, so we need to check equality of the full
+  // path too.
+  state += GetCwd();
+  state += '\0';
+  for (int i = 0; i < argc; ++i) {
+    string arg;
+    GetWin32EscapedString(argv[i], &arg);
+    state += arg;
+    state += '\0';
+  }
+  state += kNinjaVersion;
+  state += '\0';
+  // Append environment variables.
+  LPTCH env = GetEnvironmentStrings();
+  // Windows stores some stuff in private env vars starting with '=' that we
+  // want to skip. In particular, the exit code of the previous command, which
+  // will confuse us into thinking a real env var has changed.
+  LPTCH start = env;
+  while (*start == '=') {
+    while (*start) ++start;
+    ++start;
+  }
+  if (*start) {
+    LPTCH end = start;
+    // Find the end of the environment variables, marked by two consecutive null
+    // chars.
+    while (*end || *(end + 1)) ++end;
+    state.append(start, sizeof(TCHAR) * (end - start));
+  }
+  FreeEnvironmentStrings(env);
+  return state;
+}
+
+string GetPipeName() {
+  const int max_len =
+      246;  // max pipe name length - required pipe prefix length
+  string cwd = GetCwd().substr(0, max_len);
+  replace(cwd.begin(), cwd.end(), '\\', '/');
+  return R"(\\.\pipe\)" + cwd;
+}
+
+string GetEventName() {
+  string cwd = GetCwd().substr(0, MAX_PATH);
+  replace(cwd.begin(), cwd.end(), '\\', '/');
+  return cwd;
+}
+
+void StartServer(int argc, char **argv) {
+  // Assemble command line.
+  string args;
+  for (int i = 0; i < argc; i++) {
+    string escaped;
+    GetWin32EscapedString(argv[i], &escaped);
+    args += escaped;
+    args += ' ';
+  }
+  // Start the process and wait for it to create the pipe before continuing.
+  STARTUPINFO si = {0};
+  PROCESS_INFORMATION pi = {0};
+  HANDLE pipe_created_event =
+      CreateEvent(NULL, TRUE, FALSE, GetEventName().c_str());
+  if (!CreateProcess(NULL, (char *)args.c_str(), NULL, NULL, FALSE, 0, NULL,
```

Comment:
Although we are still using `CreateProcessA`, `CreateProcessW` might actually modify the second argument. Microsoft might also change the behavior. I would copy the args into a buffer instead of the `const_cast` here. (or at least use `const_cast` instead)

Reply:
Done. Hopefully CreateProcessW will never be required now that Microsoft is moving toward supporting UTF-8 as the system locale code page.